File: Java/838.java
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
    public String pushDominoes(String dominoes) {
        
        char[] dom = dominoes.toCharArray();
        
        int N = dom.length;
        for(int i=0; i< N;i++){
            
            if(dom[i] == '.') continue;
            
            if(dom[i] == 'R'){
                int count = 0,j=i+1;
              //  System.out.println("Processing R "+ i);
                while(j<N && dom[j] != 'L') {
                    if(dom[j]=='R') {
                        while(i < j) dom[i++] ='R';
                    } 
                    j++;
                    
                }
                int k = i+1;
                int p = j;
                if(j == N && dom[N-1] !='L'){
                    while(k < N) dom[k++] ='R';
                    break;
                }
                j--;
                
                while(k < j){
                  //  System.out.println("Moving ahead  "+ k + ",  " + j);
                    dom[k] = 'R';
                    dom[j] = 'L';
                    k++;
                    j--;
                }
                i=p;
            } else{
                int j = i-1;
                while(j >= 0 && dom[j] !='R' && dom[j] !='L'){
                   // System.out.println("Processing L "+ j);
                    dom[j]='L';
                    j--;
                } 
            }
            
        }
        return new String(dom);
    }
}
__________________________________________________________________________________________________
sample 38520 kb submission
class Solution {
    public String pushDominoes(String dominoes) {
        if (dominoes == null || dominoes.length() <= 1) {
            return dominoes;
        }
        int len = dominoes.length();
        int[] leftDis = new int[len];
        Arrays.fill(leftDis, Integer.MAX_VALUE);
        char prev = '.';
        int lastIndex = -1;
        for (int i = 0; i < len; i++) {
            char c = dominoes.charAt(i);
            if (c == 'R') {
                prev = 'R';
                lastIndex = i;
            } else if (c == 'L') {
                prev = '.';
            } else {
                if (prev == 'R') {
                    leftDis[i] = i - lastIndex;
                }
            }
        }
        int[] rightDis = new int[len];
        Arrays.fill(rightDis, Integer.MAX_VALUE);
        prev = '.';
        lastIndex = len;
        for (int i = len - 1; i >= 0; i--) {
            char c = dominoes.charAt(i);
            if (c == 'L') {
                prev = 'L';
                lastIndex = i;
            } else if (c == 'R') {
                prev = '.';
            } else {
                if (prev == 'L') {
                    rightDis[i] = lastIndex - i;
                }
            }
        }
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < len; i++) {
            char c = dominoes.charAt(i);
            if (c == 'L' || c == 'R') {
                res.append(c);
            } else {
                if (leftDis[i] == rightDis[i]) {
                    res.append('.');
                } else if (leftDis[i] < rightDis[i]) {
                    res.append('R');
                } else {
                    res.append('L');
                }
            }
        }
        return res.toString();
    }
}
__________________________________________________________________________________________________

File: Python3/838.py
__________________________________________________________________________________________________
sample 84 ms submission
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        res = ''
        count = 0
        for d in dominoes:
            if d == '.':
                count += 1
            elif not res or res[-1] == 'L':
                if d == 'L':
                    res += 'L' * (count + 1)
                else:
                    res += '.' * count + 'R'
                count = 0
            else:
                if d == 'R':
                    res += 'R' * (count + 1)
                else:
                    res += ('R' * (count // 2) 
                            + ('.' if count & 1 else '')
                            + 'L' * (count // 2 + 1))
                count = 0
        if count > 0:
            if not res or res[-1] == 'L':
                res += '.' * count
            elif res[-1] == 'R':
                res += 'R' * count
        return res
__________________________________________________________________________________________________
sample 88 ms submission
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        ans, cnt, toRight = '', 0, False
        for c in dominoes:
            if c == 'L':
                if toRight:
                    ans += 'R'*(cnt//2) + '.'*(cnt-cnt//2*2) + 'L'*(cnt//2+1)
                else:
                    ans += 'L'*(cnt+1)
                cnt, toRight = 0, False
            elif c == 'R':
                if toRight:
                    ans += 'R'*(cnt+1)
                else:
                    ans += '.'*cnt + 'R'
                cnt, toRight = 0, True
            else:
                cnt += 1
        if cnt > 0:
            if toRight:
                ans += 'R'*cnt
            else:
                ans += '.'*cnt
        return ans
__________________________________________________________________________________________________
